pkg/foundation/grpcutil/websocket.go

```diff
@@ -0,0 +1,253 @@
+		p.handler.ServeHTTP(response, r)
+		p.logger.Debug(ctx).Err(ctx.Err()).Msg("closing pipes")
+		responseW.CloseWithError(io.EOF)
+	}()
+
+	messages := make(chan []byte)
+	// startWebSocketRead and startWebSocketWrite need to cancel the context
+	// if they encounter an error reading from or writing to the WS connection
+	go p.startWebSocketRead(ctx, conn, cancelFn)
+	go p.readFromHTTPResponse(ctx, responseR, messages)
+	p.startWebSocketWrite(ctx, messages, conn, cancelFn)
+}
+
+// startWebSocketRead starts a read loop on the proxy's WebSocket connection.
+// The read loop will stop if there's been an error reading a message.
+func (p *webSocketProxy) startWebSocketRead(ctx context.Context, conn *websocket.Conn, onDone func()) {
+	defer onDone()
+
+	conn.SetReadLimit(512)
+	err := conn.SetReadDeadline(time.Now().Add(p.pongWait))
+	if err != nil {
+		p.logger.Warn(ctx).Err(err).Msgf("couldn't set read deadline %v", p.pongWait)
+		return
+	}
+
+	conn.SetPongHandler(func(string) error {
+		err := conn.SetReadDeadline(time.Now().Add(p.pongWait))
+		if err != nil {
+			// todo return err?
+			p.logger.Warn(ctx).Err(err).Msgf("couldn't set read deadline %v", p.pongWait)
+		}
+		return nil
+	})
+
+	for {
+		// The only use we have for reads right now
+		// is for ping, pong and close messages.
+		// https://pkg.go.dev/github.com/gorilla/websocket#hdr-Control_Messages
+		// Also, a read loop can detect client disconnects much quicker:
+		// https://groups.google.com/g/golang-nuts/c/FFzQO26jEoE/m/mYhcsK20EwAJ
+		_, _, err := conn.ReadMessage()
+		if err != nil {
+			if p.isClosedConnErr(err) {
+				p.logger.Debug(ctx).Err(err).Msg("closed connection")
+			}
+
+			p.logger.Warn(ctx).Err(err).Msg("read error")
+			break
+		}
+	}
+}
+
+func (p *webSocketProxy) isClosedConnErr(err error) bool {
+	str := err.Error()
+	if strings.Contains(str, "use of closed network connection") {
+		return true
+	}
+	return websocket.IsCloseError(
+		err,
+		websocket.CloseNormalClosure,
+		websocket.CloseGoingAway,
+		websocket.CloseAbnormalClosure,
+	)
+}
+
+func (p *webSocketProxy) readFromHTTPResponse(ctx context.Context, responseReader io.Reader, c chan []byte) {
+	defer close(c)
+	scanner := bufio.NewScanner(responseReader)
+
+	for scanner.Scan() {
+		if len(scanner.Bytes()) == 0 {
+			p.logger.Warn(ctx).Err(scanner.Err()).Msg("[write] empty scan")
+			continue
+		}
+
+		p.logger.Trace(ctx).Msgf("[write] scanned %v", scanner.Text())
+		c <- scanner.Bytes()
+	}
+
+	if sErr := scanner.Err(); sErr != nil {
+		p.logger.Err(ctx, sErr).Msg("failed reading data from original response")
+		c <- []byte(sErr.Error())
+	}
+
+	p.logger.Debug(ctx).Msg("scanner reached end of input data")
+}
+
+func (p *webSocketProxy) startWebSocketWrite(ctx context.Context, messages chan []byte, conn *websocket.Conn, cancelFn func()) {
+	defer func() {
+		for range messages {
+			// throw away
+		}
+	}()
+	defer cancelFn()
+
+	ticker := time.NewTicker(p.pingPeriod)
+	defer ticker.Stop()
```

Comment:
I know I have written this, but now that I look at it this could be nicer 🙂 
```suggestion
	ticker := time.NewTicker(p.pingPeriod)
	defer func() {
		ticker.Stop()
		cancelFn()
		for range messages {
			// throw away
		}
	}()
```

Reply:
Makes sense, having a single defer looks nicer. :+1: